DataFormats/ParticleFlowCandidate/src/PFCandidate.cc

```diff
@@ -296,8 +296,14 @@ void PFCandidate::rescaleMomentum(double rescaleFactor) {
     throw cms::Exception(
         "NegativeScaling",
         "Scale factor " + std::to_string(rescaleFactor) + " is < 0. Cannot rescale momentum by this value");
-  float rescaleE = std::sqrt(p() * p() * (rescaleFactor * rescaleFactor - 1) / (energy() * energy()) + 1);
-  LorentzVector rescaledp4(rescaleFactor * px(), rescaleFactor * py(), rescaleFactor * pz(), rescaleE * energy());
+
+  float e = std::sqrt(p() * p() * rescaleFactor * rescaleFactor + mass() * mass());
+
+  // Protect against invalid values (shouldn't happen, but could)
```

Comment:
I wonder whether with the new formulation above (no more differences that could lead to numerical instabilities) this check is still needed. Did you try in [your test](https://github.com/cms-sw/cmssw/issues/41397#issuecomment-1535909191) if `e` can get some invalid value now?

Reply:
No, I didn't see invalid values after this fix. I *think* what happened is that the scaling was applied multiple times, the first time being screwed up by the precision, and the second time giving a negative value for e^2 due to a effectively having a negative mass from the inconsistent E value. Probably this check isn't needed, or it could raise and exception instead of silently correcting the value. I'm open to suggestions.